engine/table/src/main/java/io/deephaven/engine/table/impl/AbstractFilterExecution.java

```diff
@@ -232,77 +196,338 @@ public LogOutput append(LogOutput output) {
+        final long costCeiling = filterIdx + 1 < filterContexts.length
+                ? filterContexts[filterIdx + 1].pushdownFilterCost()
+                : Long.MAX_VALUE;
+
+        // Result consumer for normal filtering.
+        final Consumer<WritableRowSet> onFilterComplete = (result) -> {
+            // Clean up the row sets created by the filter.
+            try (final RowSet ignored = localInput.getValue()) {
+                // Store the output as the next filter input.
+                localInput.setValue(result);
+            }
+            // This filter is complete, sort the remaining filters and call the consumer.
+            if (filterIdx + 1 < filterContexts.length) {
+                // Recompute the remaining costs to ensure the next filter is the lowest cost.
+                FilterContext.sortFilterContexts(
+                        filterIdx, filterContexts, sourceTable, localInput.getValue(), usePrev);
+            }
+            filterComplete.run();
+        };
+
+        // Result consumer for push-down filtering.
+        final Consumer<PushdownResult> onPushdownComplete = (pushdownResult) -> {
+            fc.updateExecutedFilterCost(costCeiling);
+
+            if (pushdownResult.maybeMatch().isEmpty()) {
+                localInput.setValue(pushdownResult.match().copy());
+
+                // Recompute the remaining costs to ensure the next filter is the lowest cost.
+                if (filterIdx + 1 < filterContexts.length) {
+                    FilterContext.sortFilterContexts(filterIdx + 1, filterContexts, sourceTable,
+                            localInput.getValue(), usePrev);
+                }
+
+                // Cleanup the rowsets and call the consumer.
+                try (final PushdownResult ignored = pushdownResult) {
+                    filterComplete.run();
+                    return;
+                }
+            }
+
+            // We still have some maybe rows, sort the filters again.
+            if (filterIdx + 1 < filterContexts.length) {
+                // Recompute the remaining costs to ensure the next filter is the lowest cost.
+                FilterContext.sortFilterContexts(
+                        filterIdx, filterContexts, sourceTable, localInput.getValue(), usePrev);
+            }
+
+            // If there is a new filter at the current index, need to evaluate it.
+            if (!fc.equals(filterContexts[filterIdx])) {
+                // Use the union of the match and maybe rows as the input for the next filter.
+                localInput.setValue(pushdownResult.match().union(pushdownResult.maybeMatch()));
+
+                // Store the result for later use by the companion chunk filter.
+                fc.updatePushdownResult(pushdownResult);
+
+                // Do the next round of filtering with the new filter that bubbled up to the current index.
+                executeStatelessFilter(
```

Comment:
After you've done a pass of pushdown; and still have a set of filters to execute, are we preserving the original order of the filters?  We should likely add some sort of test to verify that (perhaps creating a few bogus column sources that have pushdown but won't complete everything, and record when we actually read from them in order to do the final chunk filter).